pjsip/src/pjsua-lib/pjsua_call.c

```diff
@@ -4147,6 +4171,452 @@ static pj_status_t process_pending_reinvite(pjsua_call *call)
+	delay.sec = 0;
+	delay.msec = (1 << call->trickle_ice.retrans18x_count)*
+		     pjsip_cfg()->tsx.t1;
+    } else {
+	delay.sec = 1;
+	delay.msec = 500;
+    }
+    pjsua_schedule_timer(te, &delay);
+}
+
+
+static void trickle_ice_recv_sip_info(pjsua_call *call, pjsip_rx_data *rdata)
+{
+    pjsip_media_type med_type;
+    pjsip_rdata_sdp_info *sdp_info;
+    pj_status_t status;
+    unsigned i, j, med_cnt;
+    pj_bool_t use_med_prov;
+
+    pjsip_media_type_init2(&med_type, "application", "trickle-ice-sdpfrag");
+
+    /* Parse the SDP */
+    sdp_info = pjsip_rdata_get_sdp_info2(rdata, &med_type);
+    if (!sdp_info->sdp) {
+	pj_status_t err = sdp_info->body.ptr? sdp_info->sdp_err:PJ_ENOTFOUND;
+	pjsua_perror(THIS_FILE, "Failed to parse trickle ICE SDP in "
+				"incoming INFO", err);
+	return;
+    }
+
+    PJSUA_LOCK();
+
+    /* Retrieve the candidates from the SDP */
+    use_med_prov = call->med_prov_cnt > call->med_cnt;
+    med_cnt = use_med_prov? call->med_prov_cnt : call->med_cnt;
+    for (i = 0; i < sdp_info->sdp->media_count; ++i) {
+	pjmedia_transport *tp = NULL;
+	pj_str_t mid, ufrag, pwd;
+	unsigned cand_cnt = PJ_ICE_ST_MAX_CAND;
+	pj_ice_sess_cand cand[PJ_ICE_ST_MAX_CAND];
+	pj_bool_t end_of_cand;
+
+	status = pjmedia_ice_trickle_decode_sdp(sdp_info->sdp, i, &mid,
+						&ufrag, &pwd,
+						&cand_cnt, cand,
+						&end_of_cand);
+	if (status != PJ_SUCCESS) {
+	    pjsua_perror(THIS_FILE, "Failed to retrive ICE candidates from "
+				    "SDP in incoming INFO", status);
+	    continue;
+	}
+
+	for (j = 0; j < med_cnt; ++j) {
+	    pjsua_call_media *cm = use_med_prov? &call->media_prov[j] :
+						 &call->media[j];
+	    tp = cm->tp_orig;
+
+	    if (tp && tp->type == PJMEDIA_TRANSPORT_TYPE_ICE &&
+		pj_strcmp(&cm->rem_mid, &mid) == 0)
+	    {
+		break;
+	    }
+	}
+
+	if (j == med_cnt)
+	    continue;
+
+	/* Update ICE checklist */
+	status = pjmedia_ice_trickle_update(tp, &ufrag, &pwd, cand_cnt, cand,
+					    end_of_cand);
+	if (status != PJ_SUCCESS) {
+	    pjsua_perror(THIS_FILE, "Failed to update ICE checklist from "
+				    "incoming INFO", status);
+	}
+    }
+
+    PJSUA_UNLOCK();
+}
+
+
+static void trickle_ice_send_sip_info(pj_timer_heap_t *th,
+				      struct pj_timer_entry *te)
+{
+    pjsua_call *call = (pjsua_call*)te->user_data;
+    pj_pool_t *tmp_pool = NULL;
+    pj_bool_t all_end_of_cand, use_med_prov;
+    pjmedia_sdp_session *sdp;
+    unsigned i, med_cnt;
+    pjsua_msg_data msg_data;
+    pjsip_generic_string_hdr hdr1, hdr2;
+    pj_status_t status = PJ_SUCCESS;
+    pj_bool_t forced, need_send = PJ_FALSE;
+    pj_sockaddr orig_addr;
+
+    pj_str_t SIP_INFO		= {"INFO", 4};
+    pj_str_t CONTENT_DISP_STR	= {"Content-Disposition", 19};
+    pj_str_t INFO_PKG_STR	= {"Info-Package", 12};
+    pj_str_t TRICKLE_ICE_STR	= {"trickle-ice", 11};
+
+    PJ_UNUSED_ARG(th);
+
+    PJSUA_LOCK();
+
+    /* Check provisional media or active media to use */
+    use_med_prov = call->med_prov_cnt > call->med_cnt;
```

Comment:
I'm wondering if this is the right check? If media is not being added, is it correct to always use the active media?

Btw, there's already an internal function: pjsua_call_media_is_changing() for the same check.

Reply:
Actually I cannot remember exactly the lifetime (or the validity period) of the provisional media, while currently the `trickle_ice_send_sip_info()` does not acquire call/dlg lock, so its execution may be started before SDP nego but ended after SDP nego. The call media is needed only to make sure that the media transport base is ICE, changing transport base does not seem to be a common case. So active media should be the safer choice?

Or perhaps it is better to always protect the function with call/dlg lock (a bit costly considering the function will be called every 100ms until all ICE candidates are gathered) and evaluate provisional media whenever it is valid (med_prov_cnt>0?) rather than active media as generally provisional media is more updated.

Reply:
If only to check the transport type, I think we can leave it as it is now.